api/types/container/hostconfig.go

```diff
@@ -377,16 +377,17 @@ type UpdateConfig struct {
 // Portable information *should* appear in Config.
 type HostConfig struct {
 	// Applicable to all platforms
-	Binds           []string      // List of volume bindings for this container
-	ContainerIDFile string        // File (path) where the containerId is written
-	LogConfig       LogConfig     // Configuration of the logs for this container
-	NetworkMode     NetworkMode   // Network mode to use for the container
-	PortBindings    nat.PortMap   // Port mapping between the exposed port (container) and the host
-	RestartPolicy   RestartPolicy // Restart policy to be used for the container
-	AutoRemove      bool          // Automatically remove container when it exits
-	VolumeDriver    string        // Name of the volume driver used to mount volumes
-	VolumesFrom     []string      // List of volumes to take from other container
-	ConsoleSize     [2]uint       // Initial console size (height,width)
+	Binds           []string          // List of volume bindings for this container
+	ContainerIDFile string            // File (path) where the containerId is written
+	LogConfig       LogConfig         // Configuration of the logs for this container
+	NetworkMode     NetworkMode       // Network mode to use for the container
+	PortBindings    nat.PortMap       // Port mapping between the exposed port (container) and the host
+	RestartPolicy   RestartPolicy     // Restart policy to be used for the container
+	AutoRemove      bool              // Automatically remove container when it exits
+	VolumeDriver    string            // Name of the volume driver used to mount volumes
+	VolumesFrom     []string          // List of volumes to take from other container
+	ConsoleSize     [2]uint           // Initial console size (height,width)
+	OCIAnnotations  map[string]string `json:",omitempty"` // Annotations to pass along to OCI runtime
```

Comment:
Before we merge; quick check on the naming. Wondering if `OCI` prefix is needed in our API here. I get where we're coming from; mostly thinking along the lines of "(almost) every option we have in our API leads to `<something>` OCI spec.

So wondering if we should keep the abstraction - call it `Annotation`, and keep the fact that that (currently) gets passed on to the OCI to the description only.

Reply:
I'm coming around to dropping the `OCI` prefix and adopting essentially [the same semantics as Kubernetes `Annotations`.](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) And by that I mean documenting that Engine API clients can also leverage annotations to attach and consume metadata on containers for their own purposes. Perhaps we should also follow their lead and [enforce a particular syntax for annotation keys](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set) so that the namespace of an annotation is unambiguous. We could always loosen validation later but can't make it more restrictive.

Reply:
Let's discuss / brainstorm in the call later Today (I also need to get my head around all of that). Quick comment;

- Yes, I think Annotations could be useful for "our side" as well
- I like namespaces / prefixes as they can definitely help prevent ambiguity (I think I may actually have been the one originally getting them into the "labels" definition 🤔 )
- ^^^ w.r.t. namespaces: I always keep in the back of my mind that `API !== UX`; namespaces _are_ awkward to use, but are (in most cases) not meant for humans. So if we decide to use namespaces, we can still make the UX "different" (`docker run --foo=bar` doing `docker run --annotation=awkwardly-long-namespace.option=bar` behind the scenes)

Reply:
What I like about how K8s goes about namespacing annotations is that it makes namespace prefixes optional and [reserves the no-namespace "namespace" for users.](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set:~:text=If%20the%20prefix%20is%20omitted%2C%20the%20annotation%20Key%20is%20presumed%20to%20be%20private%20to%20the%20user.)

Reply:
Yup, that was part of my original suggestion for labels; no-namespace is for users; don't use them for automation as they can easily conflict. For automation / tools, always use a prefix

Reply:
For those interested; here's the PR with all the fun (well, there's been multiple before that, but that's the one that made it, and that one is already 200+ comments 🤣)

- https://github.com/moby/moby/pull/9882

Reply:
Decisions from today's maintainer call:
- Drop the `OCI` prefix from the field name
- Any annotation which is valid per the OCI Runtime spec is permitted
  - Keys can be any JSON string except for the empty string
  - Values can be any JSON string, including the empty string
  - Keys and values can be arbitrarily long